apis/supervisor/idp/v1alpha1/types_githubidentityprovider.go.tmpl

```diff
@@ -0,0 +1,257 @@
+	GitHubPhaseReady GitHubIdentityProviderPhase = "Ready"
+
+	// GitHubPhaseError is the phase for an GitHubIdentityProvider in an unhealthy state.
+	GitHubPhaseError GitHubIdentityProviderPhase = "Error"
+)
+
+type GitHubAllowedAuthOrganizationsPolicy string
+
+const (
+	// GitHubAllowedAuthOrganizationsPolicyAllGitHubUsers means any GitHub user is allowed to log in using this identity
+	// provider, regardless of their organization membership or lack thereof.
+	GitHubAllowedAuthOrganizationsPolicyAllGitHubUsers GitHubAllowedAuthOrganizationsPolicy = "AllGitHubUsers"
+
+	// GitHubAllowedAuthOrganizationsPolicyOnlyUsersFromAllowedOrganizations means only those users with membership in
+	// the listed GitHub organizations are allowed to log in.
+	GitHubAllowedAuthOrganizationsPolicyOnlyUsersFromAllowedOrganizations GitHubAllowedAuthOrganizationsPolicy = "OnlyUsersFromAllowedOrganizations"
+)
+
+// GitHubIdentityProviderStatus is the status of an GitHub identity provider.
+type GitHubIdentityProviderStatus struct {
+	// Phase summarizes the overall status of the GitHubIdentityProvider.
+	//
+	// +kubebuilder:default=Pending
+	// +kubebuilder:validation:Enum=Pending;Ready;Error
+	Phase GitHubIdentityProviderPhase `json:"phase,omitempty"`
+
+	// Conditions represents the observations of an identity provider's current state.
+	//
+	// +patchMergeKey=type
+	// +patchStrategy=merge
+	// +listType=map
+	// +listMapKey=type
+	Conditions []metav1.Condition `json:"conditions,omitempty" patchStrategy:"merge" patchMergeKey:"type"`
+}
+
+// GitHubAPIConfig allows configuration for GitHub Enterprise Server
+type GitHubAPIConfig struct {
+	// Host is required only for GitHub Enterprise Server.
+	// Defaults to using GitHub's public API (github.com).
+	// Do not specify a protocol or scheme since 'https://' will always be used.
+	//
+	// +kubebuilder:default="github.com"
+	// +kubebuilder:validation:MinLength=1
+	// +kubebuilder:validation:XValidation:message="Do not specify a URL, only a domain name",rule="!isURL(self)"
+	// +kubebuilder:validation:XValidation:message="Do not specify a path, only a domain name",rule="!isURL(self) && isURL('https://'+self) && url('https://'+self).getEscapedPath() == ''"
+	// +kubebuilder:validation:XValidation:message="Do not specify a query, only a domain name",rule="!isURL(self) && isURL('https://'+self) && size(url('https://'+self).getQuery()) < 1"
+	// +kubebuilder:validation:XValidation:message="Do not specify a fragment, only a domain name",rule="!self.contains('#')"
+	// +optional
+	Host *string `json:"host"`
+
+	// TLS configuration for GitHub Enterprise Server.
+	//
+	// +kubebuilder:validation:XValidation:message="spec.gitHubAPI.tls.certificateAuthorityData must be a PEM bundle starting with '-----BEGIN CERTIFICATE----' and then base64-encoded",rule="has(self.certificateAuthorityData) && self.certificateAuthorityData.startsWith('LS0tLS1CRUdJTiBDRVJUSUZJQ0FURS0tLS0tC')"
```

Comment:
Are we sure we want this simplistic check here? The controller could use the go std library to do much more correct checks handling all possible edge cases.

Reply:
One possible check for the controller is to see if the go library found any useful certificates in this bundle. The parsing function (I forgot the name right now) returns a boolean to indicate this.

Reply:
Let's remove it and use CRD status to check, see next push